pkg/networkservice/common/heal/client.go

```diff
@@ -19,41 +19,208 @@ package heal
 
 import (
 	"context"
+	"sync"
+	"time"
 
+	"github.com/edwarnicke/serialize"
 	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
 	"google.golang.org/grpc"
 
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
-
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
 )
 
-// RegisterClientFunc - required to inform heal server about new client connection and assign it to the connection ID
-type RegisterClientFunc func(context.Context, *networkservice.Connection, networkservice.MonitorConnectionClient)
-
 type healClient struct {
-	ctx context.Context
-	cc  networkservice.MonitorConnectionClient
+	ctx              context.Context
+	cc               networkservice.MonitorConnectionClient
+	initOnce         sync.Once
+	initErr          error
+	connCache        map[string]*networkservice.Connection
```

Comment:
Probably it would be better to create a `sync.Map` of something like that:
```go
type connnectionInfo struct {
    conn *networkservice.Connection
    state connState // { initial, ready, broken }
    updateCh chan struct{}
}
```
instead of managing multiple maps? And so it would be easier to handle different refresh/refresh on broken scenarios because there is a `state`.